modules/chemical_reactions/src/actions/ChemicalCompositionAction.C

```diff
@@ -0,0 +1,208 @@
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ChemicalCompositionAction.h"
+#include "FEProblemBase.h"
+#include "MooseMesh.h"
+#include "MooseUtils.h"
+#include "MooseUtils.h"
+
+#include "libmesh/string_to_enum.h"
+
+#ifdef THERMOCHIMICA_ENABLED
+#include "Thermochimica-cxx.h"
+#include "checkUnits.h"
+
+registerMooseAction("ChemicalReactionsApp", ChemicalCompositionAction, "add_variable");
+registerMooseAction("ChemicalReactionsApp", ChemicalCompositionAction, "add_aux_variable");
+registerMooseAction("ChemicalReactionsApp", ChemicalCompositionAction, "add_ic");
+registerMooseAction("ChemicalReactionsApp", ChemicalCompositionAction, "add_user_object");
+registerMooseAction("ChemicalReactionsApp", ChemicalCompositionAction, "add_aux_kernel");
+#endif
+
+InputParameters
+ChemicalCompositionAction::validParams()
+{
+  InputParameters params = Action::validParams();
+  params.addClassDescription("Sets up the thermodynamic model and variables for the "
+                             "thermochemistry using Thermochimica.");
+  params.addParam<std::vector<std::string>>("elements", "List of chemical elements");
+  params.addParam<FileName>("initial_values", "The CSV file name with initial conditions.");
+  params.addParam<FileName>("thermofile", "Thermodynamics model file");
+
+  MooseEnum tUnit("K C F R", "K");
+  params.addParam<MooseEnum>("tunit", tUnit, "Temperature Unit");
+  MooseEnum pUnit("atm psi bar Pa kPa", "atm");
+  params.addParam<MooseEnum>("punit", pUnit, "Pressure Unit");
+  MooseEnum mUnit(
+      "mole_fraction atom_fraction atoms moles gram-atoms mass_fraction kilograms grams pounds",
+      "moles");
+  params.addParam<MooseEnum>("munit", mUnit, "Mass Unit");
+
+  params.addParam<std::vector<std::string>>("output_phases", "List of phases to be output");
+  params.addParam<std::vector<std::string>>(
+      "output_species", "List species for which concentration in the phases is needed");
+  params.addParam<std::vector<std::string>>(
+      "element_potentials",
+      "List of chemical elements for which chemical potentials are requested");
+  return params;
+}
+
+ChemicalCompositionAction::ChemicalCompositionAction(const InputParameters & params)
+  : Action(params),
+    _elements(getParam<std::vector<std::string>>("elements")),
+    _tunit(getParam<MooseEnum>("tunit")),
+    _punit(getParam<MooseEnum>("punit")),
+    _munit(getParam<MooseEnum>("munit")),
+    _phases(getParam<std::vector<std::string>>("output_phases")),
+    _species(getParam<std::vector<std::string>>("output_species")),
+    _element_potentials(getParam<std::vector<std::string>>("element_potentials"))
+{
+  std::replace(_munit.begin(), _munit.end(), '_', ' ');
+#ifndef THERMOCHIMICA_ENABLED
+  mooseError("Thermochimica disabled");
```

Comment:
can this be created without Thermochimica?

Reply:
if so we need a real error

Reply:
No, registration is protected by `#ifdef THERMOCHIMICA_ENABLED`